FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,396 @@
+
+The order in which multiple coordinates relate to an index should matter. For example, Scikit-Learn's `BallTree` index with the Haversine metric requires providing latitude and longitude values in that specific order. As another example, the order in which levels are defined in a `pandas.MultiIndex` may affect its lexsort depth.
+
+Xarray's current data model is already based on the same index-coordinate relationships. The current data model also implies a one-to-one relationship between a dimension and an index. This one-to-one relationship works as currently it is not possible to perform label-based data selection in Xarray using multi-dimensional coordinates, but since we want to enable this feature in this proposal, we need the dimension-index relationship to evolve towards many-to-many.
+
+In the example below, we'd like to select data points based on their x, y values and/or on their latitude/longitude positions:
+
+```python
+>>> da
+<xarray.DataArray (x: 2, y: 2)>
+array([[5.4, 7.8],
+       [6.2, 4.7]])
+Coordinates:
+  * lon      (x, y) float64 10.2 15.2 12.6 17.6
+  * lat      (x, y) float64 40.2 45.6 42.2 47.6
+  * x        (x) float64 200.0 400.0
+  * y        (y) float64 800.0 1e+03
+
+>>> da.sel(lon=..., lat=...)
+>>> da.sel(x=..., y=...)
+```
+
+We would need one geographic index for the `lat` and `lon` coordinates and two indexes for the `x` and `y` coordinates, respectively.
+
+## 2. Proposed API changes
+
+### 2.1 Index wrapper classes
+
+Every index that is used to select data from Xarray objects should inherit from a base class, e.g., `XarrayIndex`, that provides some common API. `XarrayIndex` subclasses would generally consist of thin wrappers around existing index classes such as `pandas.Index`, `pandas.MultiIndex`, `scipy.spatial.KDTree`, etc.
+
+There is a variety of features that an xarray index wrapper may or may not support:
+
+- 1-dimensional vs. 2-dimensional vs. n-dimensional coordinate
+- built from a single vs multiple coordinate(s)
+- in-memory vs. out-of-core (dask) index data/coordinates (vs. other array backends)
+- range-based vs. point-wise selection
+- exact vs. inexact lookups
+
+Whether or not a `XarrayIndex` subclass supports each of the features listed above should be either declared explicitly via a common API or left to the implementation. An `XarrayIndex` subclass may encapsulate more than one underlying object used to perform the actual indexing. Such "meta" index would typically support a range of features among those mentioned above and would automatically select the optimal index object for a given indexing operation.
+
+Every `XarrayIndex` subclass must at least implement two methods:
+
+- One `build` method that takes one or more Dataset/DataArray coordinates and that returns the object(s) that will be used for the actual indexing (e.g., `pandas.Index`, `scipy.spatial.KDTree`, etc.)
```

Comment:
When would xarray use `build`? How is this different from the `XarrayIndex` constructor?

Alternatively, `XarrayIndex` could do lazy initialization to construct underlying indexes (but that could be an internal implementation detail).

Reply:
I was thinking of these two advantages of having a `build` method:

- we may want to create an `XarrayIndex` from an existing index object
- a `build` abstract method with a clear signature in the `XarrayIndex` base class might be cleaner than having to specialize `__init__` in subclasses with some implicit requirements?

This could probably makes sense for simple cases... Not for "meta" indexes, though.

More generally, I don't have a clear idea yet on what will be the API of `XarrayIndex` and I guess that it will require multiple iterations during the implementation (I could then move those `build` and `query` to the next item block in the notes).

Reply:
`build` sounds very generic, but a specialized `from_coords` classmethod for constructing an `XarrayIndex` from just the type would make a lot of sense. Not every index would need a `from_coords` method, only those you want to use in this way.

Reply:
Yes a `from_coords` class method would be nice! But I'm a bit confused, wouldn't most indexes be built from coordinates? I thought this would rather be the norm, while reusing existing index objects (I mean "underlying" index objects) and extract coordinates would be more occasional..